### DISCUSSION TITLE
Adding a hardware device in Bao

### DISCUSSION THREAD

**Discussion by maskedarray** [bao-project/bao-hypervisor#82] [General]
Hi.

I am trying to add a custom device to Bao Hypervisor which I have to read and write to. What is the best way to doing so?

Having a look at src/platform/xyz/xyz_desc.c file, it seems that the struct used is called platform. Looking into the definition of the struct platform, it seems that it does have a element for specifying devices. Specifically platform.dev_num and platform.devs elements. I do not yet have an understanding of how mapping is done for Bao's VAS itself. Would simply specifying device in this struct work.

P.S: I have been able to add a device and get it working by using mem_map_dev function and I am able to read and write to the device registers. But I am wondering if the authors implemented the functionality of adding device through config file?

**Comment by josecm** [comment]
> Looking into the definition of the struct platform, it seems that it does have a element for specifying device

I'm wondering what version/commit you are on since, currently, there are no such entries on that struct. In the past, `struct platform` and `struct vm_platform` were merged, but the device regions were used in the platform VM configuration only for assigning devices to the guest and were not used for the hypervisor platform description in `xyz_desc.c` files.

> But I am wondering if the authors implemented the functionality of adding device through config file?

Here, the answer is no. Since Bao is not supposed to manage that many devices directly, having an automatic way of mapping them does not seem that useful. The way to go about it is, as you've done, using the mem allocation and mapping APIs to map that device in its initialization function explicitly. Glad you were able to figure it out with virtually no docs (which were are working on).

BTW, we will likely add in the near-future platform-specific hooks (e.g. `platform_init` in the `init`function) so that some platform-specific drivers can be run. Maybe it will be useful for your use case.

**Comment by maskedarray** [comment]
Thank you for the detailed and prompt response. Indeed I realize now that I have been working on an earlier commit. It does make sense that hypervisor is not supposed to manage any device directly. The reason I wanted to assign a device was because I was trying to add a performance monitoring unit device to Bao and setting it up for memory bandwidth regulation of Guest VMs.

I guess I can list down the steps I followed here so anyone else looking to do same might benefit:
1. Creating a struct for the device configuration which should have attributes __packed__ and aligned to PAGE_SIZE and assigned to section .devices in linker script. 
2. Using mem_map_dev function to map the address of struct to the physical start address of the external hardware
3. Interrupts can be configured by using interrupts_reserve function and interrupts_cpu_enable function
